src/components/compliance/resources/mod.rs

```diff
@@ -0,0 +1,280 @@
+    ///
+    /// For a sink, where an external resource functions in "output" mode, this would be the
+    /// equivalent of the sink pushing its data to a network endpoint, or writing data to files,
+    /// where the external resource would be responsible for aggregating that data, or read from
+    /// those files.
+    Push,
+}
+
+/// A resource definition.
+///
+/// Resource definitions uniquely identify the resource, such as HTTP, or files, and so on. These
+/// definitions generally include the bare minimum amount of information to allow the component
+/// validation runner to create an instance of them, such as spawning an HTTP server if a source has
+/// specified an HTTP resource in the "pull" direction.
+pub enum ResourceDefinition {
+    Http(HttpConfig),
+}
+
+impl From<HttpConfig> for ResourceDefinition {
+    fn from(config: HttpConfig) -> Self {
+        Self::Http(config)
+    }
+}
+
+/// An external resource associated with a component.
+///
+/// External resources represent the hypothetical location where, depending on whether the component
+/// is a source or sink, data would be generated from or collected at. This includes things like
+/// network endpoints (raw sockets, HTTP servers, etc) as well as files on disk, and more. In other
+/// words, an external resource is a data dependency associated with the component, whether the
+/// component depends on data from the external resource, or the external resource depends on data
+/// from the component.
+///
+/// An external resource includes a direction -- push or pull -- as well as the fundamental
+/// definition of the resource, such as HTTP or file. The component type is used to further refine
+/// the direction of the resource, such that a "pull" resource used with a source implies the source
+/// will pull data from the external resource, whereas a "pull" resource used with a sink implies
+/// the external resource must pull the data from the sink.
+pub struct ExternalResource {
+    direction: ResourceDirection,
+    definition: ResourceDefinition,
+    codec: ResourceCodec,
+}
+
+impl ExternalResource {
+    /// Creates a new `ExternalResource` based on the given `direction`, `definition`, and `codec`.
+    pub fn new<D, C>(direction: ResourceDirection, definition: D, codec: C) -> Self
+    where
+        D: Into<ResourceDefinition>,
+        C: Into<ResourceCodec>,
+    {
+        Self {
+            direction,
+            definition: definition.into(),
+            codec: codec.into(),
+        }
+    }
+
+    /// Spawns this resource for use as an input to a source.
+    pub fn spawn_as_input(
+        self,
+        input_rx: mpsc::Receiver<Event>,
+        resource_coordinator: &ExternalResourceCoordinator<Configured>,
+        resource_shutdown_handle: WaitHandle,
+    ) {
+        match self.definition {
+            ResourceDefinition::Http(http_config) => http_config.spawn_as_input(
```

Comment:
I could be missing something but it seems like the match here and in `spawn_as_output` are going to contain a lot of boilerplate that is calling into the specific resource definition matched... wondering if there is a way this could be designed such as to remove that :thinking: